[TSan] Don't instrument inout accesses on known-empty types

devincoughlin opened the pull request:
Empty types (such as structs without stored properties) have a
meaningless value for their address. To avoid crashes in the Thread
Sanitizer runtime, rather than passing this unspecified value as
the address of the inout access, skip emission of the runtime call.

The bug allowing unspecified behavior here has been present since we
first added TSan support for checking Swift access races -- but codegen
changes on arm64 have recently made crashes due to the bug much more
likely.

rdar://problem/47686212

devincoughlin commented:
@swift-ci Please smoke test

yln commented on test/Sanitizers/tsan-zero-sized-struct.swift:
```diff
@@ -0,0 +1,31 @@
+// RUN: %target-swift-frontend -sanitize=thread -emit-ir -primary-file %s | %FileCheck %s
```
Can we move this "IR generation" tests into `test/SILGen/tsan_instrumentation.swift` (or move the whole test) and add a runtime test here.  I minimized the following repro from another radar.  It reproduces even on my x86_64 machine:

```
// RUN: %target-run-simple-swift(-sanitize=thread)
// REQUIRES: executable_test
// REQUIRES: tsan_runtime
// UNSUPPORTED: OS=tvos

// Ensure ThreadSanitizer does not crash for zero-sized structs.

protocol MyProtocol {
    mutating func foo()
}

struct ZeroSizedStruct: MyProtocol {
    mutating func foo() { }
}

func callFoo<R: MyProtocol>(using x: inout R) {
    x.foo()
}

func callFoo<R: MyProtocol>(_ array: [Int], using x: inout R) {
    callFoo(using: &x)
}

var x = ZeroSizedStruct()
callFoo([0], using: &x)

```

It just runs the test to make sure we don't crash (no FileCheck).

yln commented:
LGTM, one test nitpick.

Should also fix: https://bugs.swift.org/browse/SR-14012

danliew-apple commented:
LGTM aside from the comments that @yln raised. 

- [ ] Move the `tsan-zero-sized-struct.swift` test to the IRGen directory.
- [ ] Add @yln 's reproducer example in `test/Sanitizers`.

danliew-apple reviewed (changes requested):
LGTM aside from the comments that @yln raised

devincoughlin commented:
@swift-ci Please smoke test and merge

devincoughlin commented:
@swift-ci Please smoke test and merge

devincoughlin commented:
@swift-ci Please smoke test and merge

devincoughlin commented:
@swift-ci Please smoke test and merge

yln commented:
The test that I suggested still fails.  It must be a similar, but slight different case?!  Let's disable it (and track it with a new radar), so we can land the current which is an improvement (expands the set of working scenario).
```
00:50:28 Script:
00:50:28 --
00:50:28 : 'RUN: at line 1';   xcrun --toolchain default --sdk '/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.1.sdk' /Users/buildnode/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/buildbot_incremental/swift-macosx-x86_64/bin/swiftc -toolchain-stdlib-rpath -target x86_64-apple-macosx10.9  -module-cache-path /Users/buildnode/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/buildbot_incremental/swift-macosx-x86_64/swift-test-results/x86_64-apple-macosx10.9/clang-module-cache -Xlinker -rpath -Xlinker /usr/lib/swift  -Xfrontend -ignore-module-source-info  /Users/buildnode/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/swift/test/Sanitizers/tsan-zero-sized-struct.swift -g -sanitize=thread  -o /Users/buildnode/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/buildbot_incremental/swift-macosx-x86_64/test-macosx-x86_64/Sanitizers/Output/tsan-zero-sized-struct.swift.tmp_tsan-binary
00:50:28 : 'RUN: at line 2';   codesign -f -s - /Users/buildnode/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/buildbot_incremental/swift-macosx-x86_64/test-macosx-x86_64/Sanitizers/Output/tsan-zero-sized-struct.swift.tmp_tsan-binary
00:50:28 : 'RUN: at line 3';   /usr/bin/env DYLD_LIBRARY_PATH='/Users/buildnode/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/buildbot_incremental/swift-macosx-x86_64/lib/swift/macosx' LD_LIBRARY_PATH='/Users/buildnode/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/buildbot_incremental/swift-macosx-x86_64/lib/swift/macosx:' SIMCTL_CHILD_DYLD_LIBRARY_PATH='/Users/buildnode/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/buildbot_incremental/swift-macosx-x86_64/lib/swift/macosx'  /Users/buildnode/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/buildbot_incremental/swift-macosx-x86_64/test-macosx-x86_64/Sanitizers/Output/tsan-zero-sized-struct.swift.tmp_tsan-binary 2>&1
00:50:28 --
00:50:28 Exit Code: 134
00:50:28 
00:50:28 Command Output (stdout):
00:50:28 --
00:50:28 ThreadSanitizer:DEADLYSIGNAL
00:50:28 ==95194==ERROR: ThreadSanitizer: SEGV on unknown address 0x000000000000 (pc 0x00010fd5b059 bp 0x7ffedff1f6d0 sp 0x7ffedff1f680 T126119448)
00:50:28 ==95194==The signal is caused by a READ memory access.
00:50:28 ==95194==Hint: address points to the zero page.
00:50:28     #0 __tsan::MemoryAccess(__tsan::ThreadState*, unsigned long, unsigned long, int, bool, bool) tsan_rtl.cpp:876 (libclang_rt.tsan_osx_dynamic.dylib:x86_64+0x5c059)
00:50:28     #1 __tsan::ExternalAccess(void*, unsigned long, void*, void (*)(__tsan::ThreadState*, unsigned long, unsigned long, int)) tsan_external.cpp:68 (libclang_rt.tsan_osx_dynamic.dylib:x86_64+0x2a4f6)
00:50:28     #2 callFoo<A>(_:using:) tsan-zero-sized-struct.swift:29 (tsan-zero-sized-struct.swift.tmp_tsan-binary:x86_64+0x100002c4e)
00:50:28     #3 main tsan-zero-sized-struct.swift:33 (tsan-zero-sized-struct.swift.tmp_tsan-binary:x86_64+0x100002a8c)
00:50:28     #4 start <null>:2 (libdyld.dylib:x86_64+0x1acc8)
00:50:28 
00:50:28 ==95194==Register values:
00:50:28 rax = 0x00008000000062bd  rbx = 0x0000800000000000  rcx = 0x00000000000062bd  rdx = 0x0000000000000000  
00:50:28 rdi = 0x0000000110bba380  rsi = 0x0001000000000000  rbp = 0x00007ffedff1f6d0  rsp = 0x00007ffedff1f680  
00:50:28  r8 = 0x0000000000000001   r9 = 0x0000000000000000  r10 = 0x00007ffedff1f8a8  r11 = 0x000000010fd298c0  
00:50:28 r12 = 0x00000c00000062bd  r13 = 0x000000010fce2c4f  r14 = 0x0000000000000000  r15 = 0x0000000110bba380  
00:50:28 ThreadSanitizer can not provide additional info.
00:50:28 SUMMARY: ThreadSanitizer: SEGV tsan_rtl.cpp:876 in __tsan::MemoryAccess(__tsan::ThreadState*, unsigned long, unsigned long, int, bool, bool)
00:50:28 ==95194==ABORTING
00:50:28 
00:50:28 --
00:50:28 Command Output (stderr):
00:50:28 --
```

devincoughlin committed:
[TSan] Don't instrument inout accesses on known-empty types

Empty types (such as structs without stored properties) have a
meaningless value for their address. To avoid crashes in the Thread
Sanitizer runtime, rather than passing this unspecified value as
the address of the inout access, skip emission of the runtime call.

The bug allowing unspecified behavior here has been present since we
first added TSan support for checking Swift access races -- but codegen
changes on arm64 have recently made crashes due to the bug much more
likely.

rdar://problem/47686212

devincoughlin commented:
@swift-ci Please smoke test and merge

devincoughlin commented:
@swift-ci Please smoke test and merge
